query decorator (#686)
allegro/turnilo: docs/extending-turnilo.md

This file should export function named `decorator`. 
This function will be called before every Plywood query is sent to Druid from Turnilo.
Function is called with four arguments:
* Plywood query
* Request object
* Decorator options
* Plywood library instance

Decorator function should return valid Plywood expression.

```javascript
exports.decorator = function (expression, request, options, plywood) {
  const userId = request.headers["x-user-id"]; // get userId from header, you need to set this value before Turnilo
  const userColumnName = options.userColumnName; // get value from options, defined in config
  const filterClause = plywood.$(userColumnName).in([userId]); // show only rows where `userColumnName` is equal to current user id.
  return expression.substitute(e => {
    if (e instanceof plywood.RefExpression && e.name === "main") { // filter all main expression references
      e.filter(filterClause);
    }
    return null;
  });
}
```

And needed configuration:

```yaml
...
dataCubes:
    - name: cube
      queryDecorator:
        path: ./decorator.js
        options:
          userColumnName: "user_id"
```

## Plugins

> Most powerful way to extend turnilo are plugins. They are defined at top level in config and apply for whole Turnilo application.

Reviewer: If "most powerful" then put plugins at the top :)

Author: With great power comes great responsibility :)

Decorator function should return valid Plywood expression.

```javascript
exports.decorator = function (expression, request, options, plywood) {
  const userId = request.headers["x-user-id"]; // get userId from header, you need to set this value before Turnilo
  const userColumnName = options.userColumnName; // get value from options, defined in config
  const filterClause = plywood.$(userColumnName).in([userId]); // show only rows where `userColumnName` is equal to current user id.
  return expression.substitute(e => {
    if (e instanceof plywood.RefExpression && e.name === "main") { // filter all main expression references
      e.filter(filterClause);
    }
    return null;
  });
}
```

And needed configuration:

```yaml
...
dataCubes:
    - name: cube
      queryDecorator:
        path: ./decorator.js
        options:
          userColumnName: "user_id"
```

## Plugins

Most powerful way to extend turnilo are plugins. They are defined at top level in config and apply for whole Turnilo application.
You need to add your plugin as entry under `plugins` field. 
Plugin need to have two fields:
    - `name` - name for debug purposes
    - `path` - path to the js file

> And can have additional field `settings`. Content of this field would be passed to plugin so it is good place to plugin parameters.

Reviewer 2: ```suggestion
It can define additional field `settings`. Content of this field would be passed to plugin so it is good place for additional parameters.
```
